Add colored margin in Inspector for arrays and dictionaries

ajreckof opened the pull request:
Closes https://github.com/godotengine/godot-proposals/issues/2018.
Closes https://github.com/godotengine/godot-proposals/issues/6987.

This PR aims to enhance readability of nested data by adding a colored border for arrays and dictionaries as it was done for Resources in #45907. 
<img src="https://user-images.githubusercontent.com/66184050/228723001-42019e63-a19c-4cc0-98b2-c5272f51463b.png" height="180" /><img src="https://user-images.githubusercontent.com/66184050/228721922-3690237d-203c-46a8-a771-b2b83d26a096.png" height="180" />

I've updated a bit this PR to have better readability for dictionaries especially or the Panel around the element to add. Here is how it nows looks like with multiple nested data (Resources, Arrays and Dictionaries) :
<img src="https://github.com/godotengine/godot/assets/66184050/78765a10-1b08-4140-a6e8-7f2aace641c8" width = "400" /><img src="https://github.com/godotengine/godot/assets/66184050/8265f480-015c-4e70-8af5-2a387b91dd9e" width = "400" />

YuriSizov commented on editor/editor_inspector.cpp:
```diff
@@ -3525,15 +3525,16 @@ void EditorInspector::set_use_wide_editors(bool p_enable) {
 		}
-		count_subinspectors = MIN(15, count_subinspectors);
+		count_subinspectors = MIN(15, count_subinspectors - 1);
+		print_line(count_subinspectors);
```
You should remove debug prints.

YuriSizov commented on editor/editor_properties_array_dict.cpp:
```diff
@@ -424,8 +425,46 @@ void EditorPropertyArray::update_property() {
+	}
+}
+
+void EditorPropertyArray::_update_property_bg() {
+	if (!is_inside_tree()) {
```
This could potentially be a method of `EditorProperty`, so we don't have to duplicate everything. The only tricky part is the `sub_inspector`/`container` check, but that check can be extracted into a virtual method that only `EditorPropertyResource`, `EditorPropertyArray`, and `EditorPropertyDictionary` would implement.

ajreckof replied:
I succeeded in making it a method if `EditorProperty` but I had to make an adjustment. I made the sub_inspector from resource editor directly bottom_editor instead of being nested in a `VBoxContainer`. I'm not sure why it was the case but couldn't see any visual difference so I supposed it was okay to remove.

YuriSizov commented on editor/editor_properties_array_dict.cpp:
```diff
@@ -1116,7 +1197,7 @@ void EditorPropertyDictionary::update_property() {
 				PanelContainer *pc = memnew(PanelContainer);
 				property_vbox->add_child(pc);
-				pc->add_theme_style_override(SNAME("panel"), get_theme_stylebox(SNAME("DictionaryAddItem"), SNAME("EditorStyles")));
+				pc->add_theme_style_override(SNAME("panel"), get_theme_stylebox(SNAME("DictionaryAddItem")));
```
Type should remain here otherwise it doesn't work.

ajreckof replied:
On the contrary with it, would fail. The subtility here is that before `DictionaryAddItem` was a stylebox from `EditorStyles` but now there is one for each depth just as inspector background so when `_update_property_bg` is called I store (through override) which stylebox is appropriate in `DictionaryAddItem` and that is what I'm fetching here.

YuriSizov replied:
Then that's an unnecessarily complicated way to do it. Create a theme cache struct and store those items in members of that struct (look how to implement one in any control node).

YuriSizov replied:
Hmm, I guess that's how it's been implemented before. Well, can wait for a refactoring pass later then.

groud commented:
Thanks for the contribution!

I am not 100% against the idea, but I kind of find useful that only sub-resources are highlighted. It is kind of making clear which parts of the properties might be things that are not part of the current scene. Thus adding highlighting to dictionaries is a bit confusing to me, and adds a lot of noise.

I'd think I'd rather have bit more feedback on that solution before merging it, and that it might require a dedicated proposal to be sure it has enough baking.

But that being said, I don't have a better solution for now to solve the readability issue, so it might be an acceptable solution for now.

ajreckof commented:
So I've updated this PR according to https://github.com/godotengine/godot-proposals/issues/6987#issuecomment-1940142900 . I think the implementation is ot really good yet but wanted some feedback on visual aspects
I've created two editor settings : 
- an enum indicating which EditorProperty should be colored (Containers & Resources, All Resources, External Resources)
- a boolean indicating wether to add a border or not to all containers and resources even when they are not colored.

I feel like the boolean setting could be removed and always set to true but maybe some people might like it without it so I decided to keep the posibility to change for now.
Here are in the order :
- Containers & Resources
- Containers & Resources (with borders)
- All Resources
- All Resources (with borders)
- External Resources
- External Resources (with borders)

<img width="350" alt="Capture d’écran 2024-02-19 à 04 25 22" src="https://github.com/godotengine/godot/assets/66184050/6a54bc31-daed-441d-a447-c3ebdd15908d"><img width="350" alt="Capture d’écran 2024-02-19 à 04 25 22" src="https://github.com/godotengine/godot/assets/66184050/6a54bc31-daed-441d-a447-c3ebdd15908d"><img width="350" alt="Capture d’écran 2024-02-19 à 04 20 54" src="https://github.com/godotengine/godot/assets/66184050/a29a490d-0ba4-4598-841e-9e439b01746b"><img width="350" alt="Capture d’écran 2024-02-19 à 05 20 17" src="https://github.com/godotengine/godot/assets/66184050/26dde0ed-6a57-4c89-acd1-09db330819d2"><img width="350" alt="Capture d’écran 2024-02-19 à 04 28 51" src="https://github.com/godotengine/godot/assets/66184050/2da03485-bd9c-4815-b928-58341f4769b3"><img width="350" alt="Capture d’écran 2024-02-19 à 05 19 44" src="https://github.com/godotengine/godot/assets/66184050/16e581a6-99d7-4dae-9f74-97b3b60fb456">

Mickeon commented:
You could format those previews in your reply using tables https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/organizing-information-with-tables

Mickeon reviewed (commented):
Not sure about the name. This is (almost) the only thing in the entire engine called "coloration". I think the setting could actually be more obvious and be called `nested_color_mode`.

The extra `delimitate_all_container_and_resources` is definitely iffy, especially since it does nothing for one of the above setting's values. Not to mention the huge name.

My opinion is that, as is, there's little reason to want this disabled as it causes this void where some delimiters should be. Not exactly pleasing.
![image](https://github.com/godotengine/godot/assets/66727710/0a36aa7a-cc6e-4a12-8006-e43211f0fcc7)
If that is addressed, it could be part of the original setting, maybe?

ajreckof commented:
I like the `nested_color_mode` I'll fill the doc once we are fixed on an implementation

For the other parameter I'm in favor of making it go away. I kept it as it acts as it used to before this PR this is like a legacy mode. I could make it true except for the all resources and rename this mode Legacy?

ajreckof committed:
Add colored margin in Inspector for arrays and dictionaries.

Apply suggestions from code review

Co-Authored-By: A Thousand Ships <96648715+AThousandShips@users.noreply.github.com>
Co-Authored-By: Tomek <kobewi4e@gmail.com>

ajreckof commented:
I finally came back to this hopefully for the last time I reduced the size when there is no border (see comparison below). I also changed the name of the setting and added description for both settings. I'm not sure of how to integrate inside the other setting. 
<img width="350" alt="Capture d’écran 2024-02-19 à 04 20 54" src="https://github.com/godotengine/godot/assets/66184050/a29a490d-0ba4-4598-841e-9e439b01746b"><img width="410" alt="Capture d’écran 2024-04-26 à 00 07 59" src="https://github.com/godotengine/godot/assets/66184050/cd235a0b-a6c6-4b0f-86dd-0cab3a8f4d89">

AThousandShips commented on doc/classes/EditorSettings.xml:
```diff
@@ -765,6 +769,12 @@
 		</member>
+		<member name="interface/inspector/nested_color_mode" type="int" setter="" getter="">
+			Control which property editors are colored when they are opened.
+			- [b]Containers &amp; Resources:[/b] Color all Array, Dictionary and Resource Editors.
```
```suggestion
			- [b]Containers &amp; Resources:[/b] Color all Array, Dictionary, and Resource Editors.
```

AThousandShips commented on doc/classes/EditorSettings.xml:
```diff
@@ -750,6 +750,10 @@
 			[b]Note:[/b] To display translated setting names in Project Settings and Editor Settings, use [member interface/editor/localize_settings] instead.
 		</member>
+		<member name="interface/inspector/delimitate_all_container_and_resources" type="bool" setter="" getter="">
+			If [code]true[/code], add a margin around Array, Dictionary and Resource Editors that are not already colored.
```
```suggestion
			If [code]true[/code], add a margin around Array, Dictionary, and Resource Editors that are not already colored.
```

AThousandShips commented on doc/classes/EditorSettings.xml:
```diff
@@ -750,6 +750,10 @@
 		</member>
+		<member name="interface/inspector/delimitate_all_container_and_resources" type="bool" setter="" getter="">
+			If [code]true[/code], add a margin around Array, Dictionary and Resource Editors that are not already colored.
+			[b]Note:[/b] if [member interface/inspector/nested_color_mode] is set to [b]Containers &amp; Resources[/b] this parameter will have no effect since those editors will already be colored
```
```suggestion
			[b]Note:[/b] If [member interface/inspector/nested_color_mode] is set to [b]Containers &amp; Resources[/b] this parameter will have no effect since those editors will already be colored
```

AThousandShips commented on editor/editor_properties.cpp:
```diff
@@ -3442,10 +3403,28 @@ void EditorPropertyResource::fold_resource() {
+	switch (p_mode) {
+		case COLORATION_CONTAINER_RESOURCE:
+			return sub_inspector != nullptr;
+			break;
+		case COLORATION_RESOURCE:
+			return true;
+			break;
```
```suggestion
		case COLORATION_RESOURCE:
			return true;
```

AThousandShips commented on editor/editor_settings.cpp:
```diff
@@ -479,6 +479,8 @@ void EditorSettings::_load_defaults(Ref<ConfigFile> p_extra_config) {
 	EDITOR_SETTING(Variant::INT, PROPERTY_HINT_RANGE, "interface/inspector/max_array_dictionary_items_per_page", 20, "10,100,1")
 	EDITOR_SETTING(Variant::BOOL, PROPERTY_HINT_NONE, "interface/inspector/show_low_level_opentype_features", false, "")
 	EDITOR_SETTING(Variant::FLOAT, PROPERTY_HINT_RANGE, "interface/inspector/float_drag_speed", 5.0, "0.1,100,0.01")
+	EDITOR_SETTING(Variant::INT, PROPERTY_HINT_ENUM, "interface/inspector/nested_color_mode", 0, "Containers & Resources, Resources, External Resources")
```
```suggestion
	EDITOR_SETTING(Variant::INT, PROPERTY_HINT_ENUM, "interface/inspector/nested_color_mode", 0, "Containers & Resources,Resources,External Resources")
```

KoBeWi commented on editor/editor_inspector.cpp:
```diff
@@ -489,6 +489,45 @@ StringName EditorProperty::_get_revert_property() const {
+	updating_theme = true;
+
+	if (bottom_editor) {
+		ColorationMode coloration_mode = (ColorationMode)(int)EDITOR_GET("interface/inspector/coloration_mode");
```
Leftover name. I see it was renamed to `nested_color_mode`.

KoBeWi commented:
Changing color mode to Resources has no effect on built-in resources.

KoBeWi commented on editor/themes/editor_theme_manager.cpp:
```diff
@@ -2125,13 +2095,79 @@ void EditorThemeManager::_populate_editor_styles(const Ref<EditorTheme> &p_theme
 
 		p_theme->set_constant("inspector_margin", EditorStringName(Editor), 12 * EDSCALE);
 
+		// Colored EditorProperty
```
```suggestion
		// Colored EditorProperty.
```

KoBeWi commented on editor/editor_inspector.cpp:
```diff
@@ -489,6 +489,45 @@ StringName EditorProperty::_get_revert_property() const {
+		return;
+	}
+
+	updating_theme = true;
```
You can use `begin_bulk_theme_override()` instead.

KoBeWi commented on editor/editor_inspector.cpp:
```diff
@@ -489,6 +489,45 @@ StringName EditorProperty::_get_revert_property() const {
+
+	if (bottom_editor) {
+		ColorationMode coloration_mode = (ColorationMode)(int)EDITOR_GET("interface/inspector/coloration_mode");
+		bool delimitate_all_container_and_resources = (bool)EDITOR_GET("interface/inspector/delimitate_all_container_and_resources");
```
```suggestion
		bool delimitate_all_container_and_resources = EDITOR_GET("interface/inspector/delimitate_all_container_and_resources");
```
No need to cast when assigning.

ajreckof commented:
> Changing color mode to Resources has no effect on built-in resources. EDIT: Though looking at the code, it does not make sense 🤔

I inverted Resource and External Resource in the list because realized it made more sense in this order thought I had done the same inversion in the enum but in fact did not :( Fixed it.

KoBeWi commented:
External Resources coloring does not work if the external resource is inside internal resource (e.g. AtlasTexture with texture). Nor sure if intended 🤔

ajreckof commented:
> External Resources coloring does not work if the external resource is inside internal resource (e.g. AtlasTexture with texture). Nor sure if intended 🤔

This is not the expected behavior. I'll try to investigate

ajreckof commented:
<img width="408" alt="Capture d’écran 2024-05-03 à 05 47 30" src="https://github.com/godotengine/godot/assets/66184050/22e67223-9fe5-4e27-8c7e-9324dd76fe2b">
I couldn't reproduce what you described but discover the color wouldn't disappear when closing the editors which I fixed.

akien-mga commented:
The last update added a merge commit, could you rebase on `master` to remove it?

akien-mga commented:
Thanks!
